### PR TITLE
Fixes subscript 0 (&var[0]) where should use (var.data()) instead.

### PR DESCRIPTION
**Pull request by JeremyRubin**
This PR fixes up a couple places where the codebase has semi-risky uses of 0 subscript. I don't think any of these are actively bugs, but some of them are unsafe (or unexpected-errors) for unsanitized inputs (such as in addrdb or key.cpp).

This also allows us to eliminate some bounds checks, so I guess this is a net-negative patch. I don't think any of those bounds checks are performance sensitive, but worth double checking in review.

p.s. I also fix some confusing code in murmurhash (with non obvious offset-math) as it took me a little bit to see that it was correct, could separate that out if desired.

### CODE REVIEW

File: src/utilstrencodings.cpp [hunk 1]
```diff
@@ -229,7 +229,7 @@ vector<unsigned char> DecodeBase64(const char* p, bool* pfInvalid)
 string DecodeBase64(const string& str)
 {
     vector<unsigned char> vchRet = DecodeBase64(str.c_str());
-    return (vchRet.size() == 0) ? string() : string((const char*)vchRet.data(), vchRet.size());
+    return string((const char*)vchRet.data(), vchRet.size());
```
Comment (dcousens) [inline src/utilstrencodings.cpp]: `std::string` maybe?

**Review by dcousens — APPROVED** [review]

**Comment by JeremyRubin** [comment]
Sorry for the bit of line noise, there was some UB when I removed some of the branches (memset/memcpy/read/write 0 bytes on nullptr/invalid ptr) that was causing one of the builds (with more aggressive runtime checks) to fail. I added them back in and all seems well now.

**Comment by laanwj** [comment]
Whoa, good catch. 

I'm shocked that &vch[0] use was still so rampant in the code. We'd  introduced `begin_ptr` with the intention to replace those, and phased even that out again with C++11 `.data()`.

Anyhow, concept ACK.

File: src/serialize.h [hunk 2]
```diff
@@ -571,7 +571,7 @@ void Serialize_impl(Stream& os, const prevector<N, T>& v, const unsigned char&)
 {
     WriteCompactSize(os, v.size());
     if (!v.empty())
-        os.write((char*)&v[0], v.size() * sizeof(T));
+        os.write((char*) v.data(), v.size() * sizeof(T));
```
Comment (laanwj) [inline src/serialize.h]: I wonder why we're casting to `(char*)` here iso `(const char*)`, even though these are write operations and the type passed in is `const`.

There's a couple more of these. Though seems out of scope to fix in this pull.

Comment (JeremyRubin) [inline src/serialize.h]: Yeah that seems obviously incorrect, but I _think_ I understand why it was done (or at least have a theory). Because the methods are all generic w.r.t. the stream type, they probably are made to work with some arbitrary stream that can't accept a `const char *`. Hopefully not the case anymore...

Anyways, probably best to fix those up separately as you suggest.

**Comment by JeremyRubin** [comment]
There are still a few more places that were trickier/not possible to correct (where the type passed in is templated, such that it would be a pointer or a vector).

I also didn't fix anything in subdirectories, because I felt this PR was getting a bit large already, and there are a few you can still grep for pretty easily there:
 ```
❯ grep '\&v[A-z0-9_]*\[0\]' */*.cpp | nl
     1	qt/signverifymessagedialog.cpp:    ui->signatureOut_SM->setText(QString::fromStdString(EncodeBase64(&vchSig[0], vchSig.size())));
     2	rpc/misc.cpp:    return EncodeBase64(&vchSig[0], vchSig.size());
     3	script/ismine.cpp:        CRIPEMD160().Write(&vSolutions[0][0], vSolutions[0].size()).Finalize(hash.begin());
     4	script/sigcache.cpp:        CSHA256().Write(nonce.begin(), 32).Write(hash.begin(), 32).Write(&pubkey[0], pubkey.size()).Write(&vchSig[0], vchSig.size()).Finalize(entry.begin());
     5	script/sign.cpp:        CRIPEMD160().Write(&vSolutions[0][0], vSolutions[0].size()).Finalize(h160.begin());
     6	script/standard.cpp:            CHash160().Write(&vSolutions[0][0], vSolutions[0].size()).Finalize(h160);
     7	test/getarg_tests.cpp:    ParseParameters(vecChar.size(), &vecChar[0]);
     8	test/skiplist_tests.cpp:        BOOST_CHECK(vIndex[from].GetAncestor(0) == &vIndex[0]);
     9	wallet/crypter.cpp:    size_t nLen = enc.Encrypt(&vchPlaintext[0], vchPlaintext.size(), &vchCiphertext[0]);
    10	wallet/crypter.cpp:    nLen = dec.Decrypt(&vchCiphertext[0], vchCiphertext.size(), &vchPlaintext[0]);
    11	wallet/rpcwallet.cpp:    return EncodeBase64(&vchSig[0], vchSig.size());
    12	wallet/wallet.cpp:    GetStrongRandBytes(&vMasterKey[0], WALLET_CRYPTO_KEY_SIZE);
```
(there are a lot more worth checking if you don't use the hungarian notation `v` prefix as a filter)

File: src/serialize.h [hunk 3]
```diff
@@ -549,7 +549,7 @@ void Serialize(Stream& os, const std::basic_string<C>& str)
 {
     WriteCompactSize(os, str.size());
     if (!str.empty())
-        os.write((char*)&str[0], str.size() * sizeof(str[0]));
+        os.write((char*) str.data(), str.size() * sizeof(typename std::basic_string<C>::value_type));
```
Comment (TheBlueMatt) [inline src/serialize.h]: AFAICT, this is equivalent to sizeof(C) (std::basic_string<C>::value_type is std::basic_string<C, std::char_traits<C>>::value_type which is std::char_traits<C>::char_type, which is C).

File: src/utilstrencodings.cpp [hunk 4]
```diff
@@ -229,7 +229,7 @@ vector<unsigned char> DecodeBase64(const char* p, bool* pfInvalid)
 string DecodeBase64(const string& str)
 {
     vector<unsigned char> vchRet = DecodeBase64(str.c_str());
-    return (vchRet.size() == 0) ? string() : string((const char*)&vchRet[0], vchRet.size());
+    return std::string((const char*)vchRet.data(), vchRet.size());
```
Comment (TheBlueMatt) [inline src/utilstrencodings.cpp]: Hmm, based on cppreference I cant tell if this is well-defined with vchRet.size() == 0 or not.

Comment (dcousens) [inline src/utilstrencodings.cpp]: From my understanding (reading up on it),  a `len` of `0` is well-defined,  and simply an empty string.  Aka,  equivalent to `std::string()`

**Comment by gmaxwell** [comment]
Concept ACK.

File: src/serialize.h (hunk 2 above)
Comment (laanwj) [inline src/serialize.h]: Sure, but it would be wrong for a stream write to take a char* i.s.o. const char*, so fixing those just means the const-correctness spreads.

File: src/utilstrencodings.cpp (hunk 4 above)
Comment (laanwj) [inline src/utilstrencodings.cpp]: Yes, a string with size 0 is well-defined. Calling `.data()` on a string of size 0 is also defined. This should be ok.

File: src/serialize.h (hunk 3 above)
Comment (laanwj) [inline src/serialize.h]: @thebluematt is right

File: src/utilstrencodings.cpp (hunk 4 above)
Comment (sipa) [inline src/utilstrencodings.cpp]: The question is whether calling `std::string(garbage_pointer, 0)` is well-defined. I can't tell from what cppreference.com says (*The behavior is undefined if s does not point at an array of at least count elements of CharT, including the case when s is a null pointer.*).

As an alternative, what about `std::string(vchRet.begin(), vchRet.end())`?

Comment (laanwj) [inline src/utilstrencodings.cpp]: I read that as: If the count is 0, it always points at least 0 elements,  thus "point at an array of at least count elements of CharT" is always satisfied. 

Alternatively: if count is 0 it would be wrong to access the pointer's address at all.

Comment (sipa) [inline src/utilstrencodings.cpp]: I'm not convinced that "a pointer to an array of 0 elements of type T" can be any pointer.

Comment (TheBlueMatt) [inline src/utilstrencodings.cpp]: I mean when cppreference is ambiguous thats probably a good sign that trying to parse a wiki is going to get your an incorrect result...best to go to the standard if in doubt in that case. :/

File: src/serialize.h (hunk 3 above)
Comment (JeremyRubin) [inline src/serialize.h]: Yep, I was aware :) I felt that the version I chose was more similar to the existing code (the size of whatever str[0] returns), but if everyone is comfortable with sizeof(C) I can change it.

File: src/utilstrencodings.cpp (hunk 4 above)
Comment (JeremyRubin) [inline src/utilstrencodings.cpp]: I don't think it is can be a garbage pointer. vector.data() is guaranteed to return a valid range (even when empty), which I think would require that it have proper alignment.

Comment (laanwj) [inline src/utilstrencodings.cpp]: Just to be complete: this is *not* the same as `std::string(vchRet.begin(), vchRet.end())` in the general case. When creating an iterator range, if the type of vch is different from the type of string, a per-element cast will happen. So `string.size()` will be `vch.size()`. Usual issues with casting between signed and unsigned apply.

`std::string((const char*)vchRet.data(), vchRet.size())` however just return the memory area with all the elements of the vector as a string. So `string.size()` will be `vch.size()*sizeof(vch::value_type)`. It is likely also more efficient.

File: src/serialize.h (hunk 3 above)
Comment (JeremyRubin) [inline src/serialize.h]: 👍 fixed

**Comment by practicalswift** [comment]
Concept ACK! 👍

**Comment by JeremyRubin** [comment]
Rebased.

File: src/hash.cpp [hunk 5]
```diff
@@ -17,36 +17,34 @@ unsigned int MurmurHash3(unsigned int nHashSeed, const std::vector<unsigned char
...
+        h1 = ROTL32(h1, 13);
+        h1 = h1 * 5 + 0xe6546b64;
+    }
 
-        //----------
-        // tail
-        const uint8_t* tail = (const uint8_t*)(&vDataToHash[0] + nblocks * 4);
+    //----------
+    // tail
+    const uint8_t* tail = (const uint8_t*)(vDataToHash.data() + nblocks * 4);
```
Comment (ryanofsky) [inline src/hash.cpp]: In commit: "Cleanup (safe, it was checked) 0-subscript in MurmurHash3"

Could you drop this (const uint8_t*) cast here? It shouldn't be needed since we already call vDataToHash.data() with no cast above on line 27.

File: src/utilstrencodings.cpp [hunk 6]
```diff
@@ -227,8 +227,8 @@ std::vector<unsigned char> DecodeBase64(const char* p, bool* pfInvalid)
 
 std::string DecodeBase64(const std::string& str)
 {
-    std::vector<unsigned char> vchRet = DecodeBase64(str.c_str());
-    return (vchRet.size() == 0) ? std::string() : std::string((const char*)&vchRet[0], vchRet.size());
+    vector<unsigned char> vchRet = DecodeBase64(str.c_str());
```
Comment (ryanofsky) [inline src/utilstrencodings.cpp]: In commit "Fix 0-subscript in utilstrencodings.cpp"

Should put back std:: prefixes.

File: src/utilstrencodings.cpp [hunk 7]
```diff
@@ -414,8 +414,13 @@ std::vector<unsigned char> DecodeBase32(const char* p, bool* pfInvalid)
 
 std::string DecodeBase32(const std::string& str)
 {
+<<<<<<< 1f0401c11315b5090cfe2b37f6338205e2984372
```
Comment (ryanofsky) [inline src/utilstrencodings.cpp]: In commit "Fix 0-subscript in utilstrencodings.cpp"

Bad merge here.

File: src/streams.h [hunk 8]
```diff
@@ -385,7 +385,7 @@ class CDataStream
     {
         // Special case: stream << stream concatenates like stream += stream
         if (!vch.empty())
-            s.write((char*)&vch[0], vch.size() * sizeof(vch[0]));
+            s.write((char*) vch.data(), vch.size() * sizeof(typename decltype(vch)::value_type));
```
Comment (ryanofsky) [inline src/streams.h]: In commit "Fix subscript-0 in streams.h"

It would seem more direct (and shorter) to write sizeof(*vch.data()).

File: src/utilstrencodings.cpp [hunk 9]
```diff
@@ -227,8 +227,8 @@ std::vector<unsigned char> DecodeBase64(const char* p, bool* pfInvalid)
 
 std::string DecodeBase64(const std::string& str)
 {
-    vector<unsigned char> vchRet = DecodeBase64(str.c_str());
-    return (vchRet.size() == 0) ? string() : string((const char*)vchRet.data(), vchRet.size());
+    std::vector<unsigned char> vchRet = DecodeBase64(str.c_str());
```
Comment (ryanofsky) [inline src/utilstrencodings.cpp]: In commit "Remove uneccessary branches in utilstrencodings string constructors."

Some of the changes in this commit should be squashed into previous commit "In commit "Fix 0-subscript in utilstrencodings.cpp."

Also missing an n in unnecessary.

**Comment by JeremyRubin** [comment]
@ryanofsky feedback addressed.

**Review by ryanofsky — COMMENTED** [review]
utACK 275caff431cea65aa77d6a3035597b4b2b9b0052. Only changes since previous review (69fd8ab0a6543dace4b080166b21b2aafdd2c0a9) were addressing feedback and combining / rearranging some of the commits.

**Comment by laanwj** [comment]
Needs rebase again (sorry).

**Comment by JeremyRubin** [comment]
@laanwj rebased.

The addrdb commit went away because of https://github.com/bitcoin/bitcoin/commit/cf68a488a4cc78d711ca6e5e3236c6d89d689079

**Comment by MeshCollider** [comment]
Looks good to me, utACK

**Review by ryanofsky — COMMENTED** [review]
utACK 30ac7688e398bbe5400a48ecf0726b679ffb845d. Changes since previous review were addrdb changes going away and some rearranged commits.

**Comment by sipa** [comment]
utACK 30ac7688e398bbe5400a48ecf0726b679ffb845d

**Comment by TheBlueMatt** [comment]
utACK 30ac7688e398bbe5400a48ecf0726b679ffb845d

**Comment by sipa** [comment]
utACK 30ac7688e398bbe5400a48ecf0726b679ffb845d
